octodns/provider/rackspace.py

```diff
@@ -0,0 +1,385 @@
+
+        if resp_data:
+            records = self._group_records(resp_data)
+            for record_type, records_of_type in records.items():
+                for raw_record_name, record_set in records_of_type.items():
+                    data_for = getattr(self,
+                                       '_data_for_{}'.format(record_type))
+                    record_name = zone.hostname_from_fqdn(raw_record_name)
+                    record = Record.new(zone, record_name,
+                                        data_for(record_set),
+                                        source=self)
+                    zone.add_record(record)
+
+        self.log.info('populate:   found %s records',
+                      len(zone.records) - before)
+
+    def _group_records(self, all_records):
+        records = defaultdict(lambda: defaultdict(list))
+        for record in all_records:
+            self._id_map[self._key_for_record(record)] = record['id']
+            records[record['type']][record['name']].append(record)
+        return records
+
+    @staticmethod
+    def _record_for_single(record, value):
+        return {
+            'name': remove_trailing_dot(record.fqdn),
+            'type': record._type,
+            'data': value,
+            'ttl': max(record.ttl, 300),
+        }
+
+    _record_for_A = _record_for_single
+    _record_for_AAAA = _record_for_single
+
+    @staticmethod
+    def _record_for_named(record, value):
+        return {
+            'name': remove_trailing_dot(record.fqdn),
+            'type': record._type,
+            'data': remove_trailing_dot(value),
+            'ttl': max(record.ttl, 300),
+        }
+
+    _record_for_NS = _record_for_named
+    _record_for_ALIAS = _record_for_named
+    _record_for_CNAME = _record_for_named
+    _record_for_PTR = _record_for_named
+
+    @staticmethod
+    def _record_for_textual(record, value):
+        return {
+            'name': remove_trailing_dot(record.fqdn),
+            'type': record._type,
+            'data': unescape_semicolon(value),
+            'ttl': max(record.ttl, 300),
+        }
+
+    _record_for_SPF = _record_for_textual
+    _record_for_TXT = _record_for_textual
+
+    @staticmethod
+    def _record_for_MX(record, value):
+        return {
+            'name': remove_trailing_dot(record.fqdn),
+            'type': record._type,
+            'data': remove_trailing_dot(value.exchange),
+            'ttl': max(record.ttl, 300),
+            'priority': value.preference
+        }
+
+    def _get_values(self, record):
+        try:
+            return record.values
+        except AttributeError:
+            return [record.value]
+
+    def _mod_Create(self, change):
+        return self._create_given_change_values(change,
+                                                self._get_values(change.new))
+
+    def _create_given_change_values(self, change, values):
+        out = []
+        for value in values:
+            transformer = getattr(self,
```

Comment:
Not a huge deal, but this line can come out of the for loop so that it only happens once for each change.